Adding support keySimulator in remotecontrollers

tamilarasi-t12 opened the pull request:
Added the KeySimulator feature as a new module within the existing Python Raft framework remote controller module.

closes https://github.com/rdkcentral/python_raft/issues/119

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/rdkcentral/python_raft?pullRequest=120) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/rdkcentral/python_raft?pullRequest=120) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/rdkcentral/python_raft?pullRequest=120) it.</sub>

rdkcmf-jenkins commented on framework/core/remoteControllerModules/keySimulator.py:
```diff
@@ -0,0 +1,59 @@
+            if keyword and keyword not in output:
+                result = True
+        else:
+            time.sleep(delay)
```
Medium impact. Quality issue.
Control flow issues. Logically dead code.
Code can never be reached because of a logical contradiction
The indicated dead code may have performed some action; that action will never occur.

tamilarasi-t12 replied:
Resolved the control flow issue. Changed the code

Ulrond commented:
rejected, this not a branch on the code.

Ulrond commented on framework/core/remoteControllerModules/keySimulator.py:
```diff
@@ -0,0 +1,86 @@
+
+        # Initialize SSH session
+        self.session = sshConsole(
+            address=self.remoteConfig.get("ip"),
```
need to extend the example with your requirements so it's there in the example list

https://github.com/rdkcentral/python_raft/blob/master/examples/configs/example_rack_config.yml

```yaml
                    remoteController:
                        # [ remoteController: ]
                        # [ type: "olimex" ip: "192.168.0.17" port: 7 map: "llama_rc6", config: "remote_commander.yml" ]
                        # [ type: "skyProc" map: "skyq_map", config: "remote_commander.yml" ]
                        # [ type: "None" ]
                        type: "none"
                        map: "skyq_map"
                        config: "example_remote_commander.yml"
```

tamilarasi-t12 replied:
Added the keySimulator in the example list

Ulrond reviewed (changes requested):
looks ok except for that minor change to upgrade the examples

tamilarasi-t12 commented:
We forked the repository and created a branch because we do not have the option to create a branch directly in the original repository.

Ulrond commented on examples/configs/example_rack_config.yml:
```diff
@@ -71,6 +71,7 @@ rackConfig:
                         # [ type: "olimex", ip: "192.168.0.17", port: 7, map: "llama_rc6", config: "remote_commander.yml" ]
                         # [ type: "skyProc", map: "skyq_map", config: "remote_commander.yml" ]
                         # [ type: "None" ]
+                        # [ type: "keySimulator", ip: "192.168.50.99", port: 10022, username: "root", password: '', map: "keysimulator_xione", config: "keymap.yml" ]
```
you can't have any of our product names in any of the open-source and anyway it shouldn't be specific to any product, its just keymaps choices something like `rdk_keymap.yaml` But where is this python file your talking about? can you send that also for review... Should it be committed to the examples directory also? can you share offline.

tamilarasi-t12 replied:
The file name has been updated to rdk_keymap.yml. We will be adding rdk_keymap.yml to the L4_SmokeTest private repository. I am sharing the rdk_keymap.yml file through Teams for review.

Ulrond reviewed (changes requested):
there's an assumption likely due to incorrect documentation that sendKey takes in a dictionary, it does not... it takes in a rcCode from rdCode.py, and that's used in all the translation tables and all the code.

Your changes assume that a key dict goes in and therefore will not correctly translate from rcCodes to the dictiomnary.

I've raised a issue to make the code clear,  but I suspect your code change will break raft users.

https://github.com/rdkcentral/python_raft/issues/122

To perform any changes to raft, we have to ensure we test raft with raft tests, in this case there's tests here -> https://github.com/rdkcentral/python_raft/blob/master/tests/remoteController_test.py

Had these been used to test raft, it may have been noticed that rcCodes are used as the input to the commonRemote, but any changes to the core will require upgrades to the basic tests to ensure that the changes comply with the requirements of the core.

tamilarasi-t12 commented:
### Reason for removing the below code due to the Issue1 and Issue 2 mentioned in the  below observation.
 ```
 keyDictionary = {}
        for key, val in config.items():
            if isinstance(val, dict):
                for k, v in val.items():
                    keyDictionary[k] = v 
            else:
                keyDictionary[key] = val
        return keyDictionary
```

### Observation
I have cloned the python-raft repository and followed the steps to run the remoteController_test.

- Added the KeySimulator remote feature within the RemoteController module 
    File Name: Added keySimulator.py
    File Name: Updated the commonRemote.py

- Updated the rack configuration's remote controller section to include the keySimulator type.
   
```
     remoteController:
                          # [ remoteController: ]
                          # [ type: "olimex" ip: "192.168.0.17" port: 7 map: "llama_rc6", config: "remote_commander.yml" ]
                          # [ type: "skyProc" map: "skyq_map", config: "remote_commander.yml" ]
                          # [ type: "None" ]
                          type: "keySimulator"
                          map: "keysimulator_xione"
                          ip: "192.168.50.182" 
                          port: 10022
                          username: "root"
                          password: ''
                          config: "./examples/configs/keymap.yml"
```


Changes in commonRemote.py

 ```
elif self.type == "keySimulator":
          self.remoteController = keySimulator (self.log, remoteConfig)
```

**Issue 1:** 
    The keymap configuration file is not found in the Repository.
    
![image](https://github.com/user-attachments/assets/060dad35-5c94-477d-ac0c-57b36e56cd53)

**Solution:**
Added keymap.yml as per the documentation ( https://github.com/rdkcentral/python_raft/wiki/Common-Remote ).

```
remoteMaps:
  remoteCommanderMap0:
    name: "keysimulator_xione"
    prefix: "keySimulator -k"
    codes:
      POWER: "power"
      HOME: "home"
      GUIDE: "guide"
      SKY: "sky"
      NUM_0: "0"
```


**Issue 2:**
After adding keymap.yml, the below error message was observed.
    
![image](https://github.com/user-attachments/assets/5c9e6d45-6302-4112-8d34-1b14b6d3fdab)
![image](https://github.com/user-attachments/assets/329e3380-eeb3-4082-af92-aeb7f4b50c2d)

**Note:**
Are there any missing component changes or pending code updates relevant to remotecontroller components in the Python Raft repository?

**Solution 1:**
Changes are already there in the Pull request.
**Solution 2:**
Even if we make just this one change, everything will work fine.

```
        keyDictionary = {}
        for key, val in config.items():
            if isinstance(val, dict):
                for k, v in val.items():
                    keyDictionary[k] = v 
            else:
                keyDictionary[key] = val
        #return keyDictionary
        return keyDictionary.get("remoteMaps")
```


We are proposing the solution 2 to make the minimal changes in the Python Raft Repository.

### For answering this query:
**Your changes assume that a key dict goes in and therefore will not correctly translate from rcCodes to the dictionary.**

When running remoteController_test.py, the function getMappedKey retrieves keys from rcCodes.py and converts them using keymap.yml. Our recent code changes ensure that remoteController_test.py is functioning correctly.

Ulrond commented on framework/core/remoteControllerModules/keySimulator.py:
```diff
@@ -0,0 +1,86 @@
+        )
+
+
+    def sendKey(self, key: str, repeat: int , delay: int ):
```
can you correct this `key: rcCode`

tamilarasi-t12 replied:
Updated the code.

TB-1993 commented on framework/core/remoteControllerModules/keySimulator.py:
```diff
@@ -0,0 +1,87 @@
+            bool: Result of the command verification.
+        """
+        result = False
+        keyword = "term start init 1"
```
This and the check for it at the bottom is the only bit that's confusing me. Is there no better way to confirm the keys sent successfully?

tamilarasi-t12 commented on framework/core/remoteControllerModules/keySimulator.py:
```diff
@@ -0,0 +1,87 @@
+        """Send a key command with specified repeats and interval.
+
+        Args:
+            key (rc): The key to send.
```
Reverted the changes from rcCodes to string datatype.

TB-1993 reviewed (approved):
Looks good to me.

tamilarasi-t12 commented:
@TB-1993, @Ulrond  I updated the code as per our previous discussion. I made the changes according to the rdk_keymap.yml design in commonRemote.py and tested it. Could you please look at it once?

TB-1993 commented on framework/core/remoteControllerModules/keySimulator.py:
```diff
@@ -0,0 +1,86 @@
+            self.session.write(f"{key}")
+            time.sleep(delay)
+
+        # Read output after sending keys
+        output = self.session.read_until(self.prompt)
+
+        # Check for the presence of a keyword in the output
+        if keyword in output:
+            result = True
```
This check should move into the repeat loops so we can confirm the key send was repeated correctly.

tamilarasi-t12 replied:
Updated the changes

TB-1993 commented on framework/core/remoteControllerModules/keySimulator.py:
```diff
@@ -0,0 +1,86 @@
+        """
+        self.log = log
+        self.remoteConfig = remoteConfig
+        self.prompt = r"\$ "
```
This should come from the remoteConfig too. This should be configurable based on the platform, it won't always be the same.

TB-1993 commented on framework/core/remoteControllerModules/keySimulator.py:
```diff
@@ -0,0 +1,86 @@
+            username=self.remoteConfig.get("username"),
+            password=self.remoteConfig.get("password"),
+            known_hosts=self.remoteConfig.get("known_hosts"),
+            port=int(self.remoteConfig.get("port")),
```
The prompt should be passed straight into the sshConsole here `prompt=remoteConfig.get("prompt")`

tamilarasi-t12 replied:
Actually, we are taking sshConsole from this file https://github.com/rdkcentral/python_raft/blob/master/framework/core/commandModules/sshConsole.py we are not having this parameter 
**prompt=remoteConfig.get("prompt")**. There is parameter mismatch here.

tamilarasi-t12 replied:
Updated the code changes as per the comments.

TB-1993 commented on framework/core/remoteControllerModules/keySimulator.py:
```diff
@@ -0,0 +1,86 @@
+
+        # Send the key command
+        for _ in range(repeat):
+            self.session.write(f"{key}")
+            time.sleep(delay)
+
+        # Read output after sending keys
+        output = self.session.read_until(self.prompt)
+
+        # Check for the presence of a keyword in the output
+        if keyword in output:
+            result = True
```
Rather than doing any of this, we can just do `self.session.write(f"{key}", wait_for_prompt=true)`. This means the session will check for the prompt before attempting to write the session.

We don't need to do the other check as this is intended to be fire and forget, so we don't need to check if we actually ran the command correctly as it'll be obvious when they don't.

tamilarasi-t12 replied:
We have a parameter mismatch here as well.

tamilarasi-t12 replied:
Found the method. Updated the code changes.

TB-1993 reviewed (approved):
Ok lets merge this, it still won't work as commonRemote is broken, but I already have a commit with the fix, I just don't have permissions to push it to your fork.
